lib/endpoint.js

```diff
@@ -90,17 +90,89 @@ VertoEndpoint.prototype.attemptInvite = function(vertoCall, matrixSide, force) {
-        // got enough candidates when SDP has a srflx or relay candidate
-        if (c.candidate.indexOf("typ srflx") !== -1 ||
-                c.candidate.indexOf("typ relay") !== -1) {
-            enoughCandidates = true;
-            console.log("Gathered enough candidates for %s", matrixSide.mxCallId);
-            break; // bail early
+
+        var mLineIndex;
+        if (c.sdpMLineIndex !== undefined) {
+            mLineIndex = c.sdpMLineIndex;
+        }
+        else {
+            // hack: synthesise an mLineIndex based on heuristics
+            if (c.sdpMid === 'audio') {
+                mLineIndex = 0;
+            }
+            else if (c.sdpMid === 'video') {
+                mLineIndex = 1;
+            }
+            else {
+                mLineIndex = 0; // this is probably wrong though.
+            }
+        }
+        stats[mLineIndex] = stats[mLineIndex] || {};
+        if (c.candidate.indexOf("typ host") !== -1) {
+            stats[mLineIndex].hasHost = 1;
+        }
+        if (c.candidate.indexOf("typ srflx") !== -1) {
+            stats[mLineIndex].hasSrflx = 1;
+        }
+        if (c.candidate.indexOf("typ relay") !== -1) {
+            stats[mLineIndex].hasRelay = 1;   
         }
+        var m = c.candidate.match(/^candidate:(.*?) (.*?) /);
+        if (m) {
+            stats[mLineIndex].componentIdSumByFoundation =
+                stats[mLineIndex].componentIdSumByFoundation || {};
+            stats[mLineIndex].componentIdSumByFoundation[m[1]] += m[2];
+        }
+        else {
+            console.error("Can't parse candidate: " + c.candidate);
+        }
+    }
+
+    var enoughCandidates = true; // start off optimistic...
+    stats.forEach(function(stat) {
+        // we've gathered enough candidates when've received srflx or relay candidates
+        //
+        // this works because if we've got a srflx, we will always be able to use it
+        // given we know that the freeswitch isn't NATted - so if a NATted client
+        // is able to correctly discover a srflx candidate, it should also be able to
+        // talk through to FS.
+        //
+        // Similarly, if we've seen a relay candidate (which are typically the slowest
+        // to set up and last to arrive), we know that will always work too, so we don't
+        // need to wait for any others.
+
+        if (!stat.hasHost) enoughCandidates = false;
+        if (!stat.hasSrflx && !stat.hasRelay) enoughCandidates = false;
+
+        // If we see gaps in the candidate component-id sequences then we should
+        // definitely hold off until we receive the missing candidates, otherwise
+        // freeswitch has been known to segfault
+
+        Object.keys(stat.componentIdSumByFoundation).forEach(function(foundation) {
+            // sum must be a triangular number
+            var sum = stat.componentIdSumByFoundation[foundation];
+            if (sum != 1 && sum != 3) enoughCandidates = false;
```

Comment:
See earlier comment: `sum` is probably going to be a `string` so this check will never pass.